core/modules/replica/DatabaseMySQL.h

```diff
@@ -0,0 +1,975 @@
+ * set the value returned for the specified column if the value was not 'NULL'.
+ * They will return 'false' otherwise. All methods have two parameyters:
+ *
+ *   columnName - the name of a column
+ *   value      - the value (of a type which depends on the method signature)
+ *                to be initialized upon the succesful completion of a method
+ *
+ * Methods may also throw the following exceptions:
+ *
+ *   std::invalid_argument - for unknown column names
+ *   InvalidTypeError      - when the conversion of row data into a value of
+ *                           the requested type is not possible.
+ *
+ * @note the validty of each object of this class is limited by the life
+ * span of the database Connection object and a result set of the last
+ * query. Use this object only for short periods of time while iterating over
+ * a result set after each successful invocation of the iterator method
+ * Connection::next().
+ *
+ * @see Connection::next()
+ *
+ */
+class Row {
+
+public:
+
+    /// Class Connection is allowed to initialize the valid content
+    /// of rows
+    friend class Connection;
+
+    /**
+     * The class encapsulate a raw data pointer and the number of bytes
+     * in each column.
+     */
+    typedef std::pair<char const*, size_t> Cell;
+
+    /**
+     * The default constructor will initialize invalid instances of the class.
+     *
+     * @note any attempts to call most (but 'isValid', copy constructor,
+     * assignment operator and desturctor) methods of objects constracted
+     * with this state will throw exception std::logic_error. So, make sure
+     * the object is properly set by passing it for initialization to method
+     * Connection::next() when iterating over a result set.
+     *
+     * @see Connection::next()
+     */
+    Row();
+
+    /// Copy constructor
+    Row(Row const& rhs) = default;
+
+    /// The Assignment operator
+    Row& operator=(Row const& rhs) = default;
+
+    /// Destructor
+    ~Row() = default;
+
+    /// @return 'true' of the object has meaningful content
+    bool isValid() const { return _isValid; }
+
+    /// @return width of the row
+    size_t numColumns() const;
+
+    // These methods will return 'true' if the specified field is NULL
+
+    bool isNull(size_t             columnIdx)  const;
+    bool isNull(std::string const& columnName) const;
+
+    // Type-specific data extractors/converters for values of fields.
+    // There are two ways to access the values: either by a relative
+    // index of a column in a result set, or by the name of the column.
+    // The second method has some extra (though, minor) overhead.
+    //
+    // @see class Row
+
+    bool get(size_t      columnIdx,         std::string& value) const;
+    bool get(std::string const& columnName, std::string& value) const;
+
+    // Unsigned integer types
+
+    bool get(size_t columnIdx, uint64_t& value) const;
```

Comment:
I thought about this. And a reason why I chose to go with the method overloading rather than a generic template (plus optional specializations in the CC file) is that I felt the overloaded functions will work better than C++ template type inference. At least, this seemed like a better option for a small collection of types I care about.

Reply:
Besides, my goal was to build a convenient and functionally sufficient API to MySQL rather than some generic API which would be used by anyone else :-)